core/lib/multivm/src/versions/vm_latest/bootloader/utils.rs

```diff
@@ -92,48 +99,127 @@ pub(super) fn apply_tx_to_memory(
+    let block_position = get_tx_operator_l2_block_info_offset(config.subversion)
+        + config.tx_index * TX_OPERATOR_SLOTS_PER_L2_BLOCK_INFO;
+
+    memory.push((block_position, bootloader_l2_block.number.into()));
+    memory.push((block_position + 1, bootloader_l2_block.timestamp.into()));
+    memory.push((
+        block_position + 2,
+        h256_to_u256(bootloader_l2_block.prev_block_hash),
+    ));
+    memory.push((
+        block_position + 3,
+        if config.start_new_l2_block {
+            bootloader_l2_block.max_virtual_blocks_to_create.into()
+        } else {
+            U256::zero()
+        },
+    ));
+
+    if config.subversion != MultiVmSubversion::Interop
+        || config.interop_root_application_config.is_none()
+    {
+        return 0;
+    }
+
+    let interop_root_application_config = config.interop_root_application_config.unwrap();
+    for (offset, interop_root) in bootloader_l2_block.interop_roots.iter().enumerate() {
+        apply_interop_root(
+            memory,
+            interop_root_application_config.number_of_applied_interop_roots + offset,
+            interop_root.clone(),
+            config.subversion,
+            bootloader_l2_block.number,
+        );
+    }
+
+    if !config.start_new_l2_block {
+        return bootloader_l2_block.interop_roots.len();
+    }
+
+    apply_interop_root_number_in_block_number(
+        memory,
+        config.subversion,
+        bootloader_l2_block.interop_roots.len(),
+        interop_root_application_config.preexisting_blocks_number,
+    );
+
+    bootloader_l2_block.interop_roots.len()
+}
+
+fn apply_interop_root(
+    memory: &mut BootloaderMemory,
+    interop_root_offset: usize,
+    interop_root: InteropRoot,
+    subversion: MultiVmSubversion,
+    l2_block_number: u32,
+) {
+    let interop_root_slot = get_interop_root_offset(subversion);
+    // Convert the byte array into U256 words
+    let mut u256_words: Vec<U256> = vec![
+        U256::from(l2_block_number),
+        U256::from(interop_root.chain_id.as_u64()),
+        U256::from(interop_root.block_number),
+        U256::from(interop_root.sides.len()),
+    ];
+
+    u256_words.extend(
+        interop_root
+            .sides
+            .iter()
+            .map(|hash| U256::from(hash.as_bytes())),
+    );
+    let start_slot = interop_root_slot + interop_root_offset * INTEROP_ROOT_SLOTS_SIZE;
+    memory.extend((start_slot..).zip(u256_words));
+}
+
+fn apply_interop_root_number_in_block_number(
+    memory: &mut BootloaderMemory,
+    subversion: MultiVmSubversion,
+    number_of_interop_roots: usize,
+    preexisting_blocks_number: usize,
+) {
+    let first_empty_slot = get_interop_blocks_begin_offset(subversion) + preexisting_blocks_number;
+    let number_of_interop_roots_plus_one: U256 = (number_of_interop_roots + 1).into();
+    memory.push((first_empty_slot, number_of_interop_roots_plus_one));
+    memory.push((
+        get_current_number_of_roots_in_block_offset(subversion),
```

Comment:
tbh memory layout looks off. `get_current_number_of_roots_in_block_offset` returns constant value so you write to the same slot every block which is incorrect